sync/server.md

```diff
@@ -0,0 +1,404 @@
+
+* Database Server Host
+* Database Server Port
+* Database Name
+* Database Username
+* Database Password
+
+Table names will not be configured as part of the Data Source for Postgres.
+Instead, the table will be determined by the SQL query defined in the Resolver Request Mapping.
+
+### Resolver Request Mapping
+
+This is the defintion of an SQL query to execute.
+Any parameters defined in the Data Schema for the Query or Mutation being executed will be made available on the `conext` object as `context.arguments`.
+For example, to find a single record by id, use the following:
+
+```
+// readNote(id: String): Note
+
+SELECT id, title, content, timestamp FROM \"Notes\" WHERE id='{{context.arguments.id}}';
+```
+
+To update a row, use the following:
+
+```
+// updateNote(id: String,title: String): Note
+
+UPDATE \"Notes\" SET \"title\"='{{context.arguments.title}}' WHERE \"id\"='{{context.arguments.id}}' RETURNING *;
+```
+
+### Resolver Response Mapping
+
+This is the defintion for how to format the result of a SQL query to match the expected GraphQL Data Schema for that Query or Mutation.
+The result of the SQL query will be made available on the `context.result` object.
+The `result` field will always be an `array`.
+If a list of objects is expected in the Schema rsponse, this can be mapped as follows:
+
+```
+{{toJSON context.result}}
+```
+
+If a max of 1 result is expected, e.g. finding a record by id, this can be mapped as follows:
+
+```
+{{toJSON context.result.[0]}}
+```
+
+## HTTP Connector
+
+### Data Source
+
+The defintion of a HTTP Data Source will have the following parameters:
+
+* Protocol i.e. HTTP or HTTPS
+* Host e.g. www.example.com
+* Port
+* Base path e.g. /myapi/v1
+* Headers (optional)
+* Basic Auth Username (optional)
+* Basic Auth Password (optional)
+
+### Resolver Request Mapping
+
+This is the defintion of a HTTP request to make for a specific Query or Mutation.
+It will have the following parameters in conjunciton with the Data Source parameters to make a request.
+
+* Path (Added to the end of the Base path defined in the Data Source. Can include query parameters)
+* Additional Headers (optional)
+* Method e.g. GET, POST, PUT
+* Request Body (Optional)
+
+These will be defined as a JSON object, but will be treated as a string.
+This means the below defintion would actually be stored as the string beneath
+
+JSON version
+```
+{
+  "path": "/notes",
+  "method": "GET",
+  "headers": {
+    "accept": "application/json"
+  }
+}
+```
+
+String version
+```
+{\n  \"path\": \"/notes\",\n  \"method\": \"GET\",\n  \"headers\": {\n    \"accept\": \"application/json\"\n  }}
+```
+
+By treating the defintion as a string, it allows for Handlebars syntax to be added anywhere in the JSON object.
+For example, the below JSON defintion could be used for a simple find Query.
+
+```
+{
+  "path": "/notes?id={{context.arguments.id}}",
+  "method": "GET",
+  "headers": {
+    "accept": "application/json"
+  }
+}
+```
```

Comment:
is it possible to use websockets in this http resolver?

Reply:
I would see websockets being a separate data source

Reply:
so how does subscription work for http resolvers?

Reply:
If you take a look at the example in this readme
https://github.com/apollographql/graphql-subscriptions#getting-started-with-your-first-subscription

and another example in this simple server
https://github.com/david-martin/starwars-server/commit/af5804d177700b6e1fd0b79fe2bc3967cd5f6861

In GraphQL, a subscription can notify clients whenever another client has executed a query or mutation.
The subscription is not at the database/data source level, it's within the graphql server as requests pass through it.

So for http, this is the flow:

* client sends a Mutation (createNote) to create a Note object
* the server executes the appropriate resolver
* in that resolver, an event is published that this mutation was executed (successfully)
* separately another client connects to the server over a websocket, and is subscribed to a Subscription (noteCreated)
* in that subscription resolver on the server, it is subscribed to any 'createNote' events
* as it hears about any 'createNote' event, it notifes any connected clients that are subscribe to this, and can include some payload (depending on what the schema defines)

More info here https://www.apollographql.com/docs/apollo-server/v2/features/subscriptions.html